File: AIM/AIM/Domain/Entities/Track.m
//
//  Track.m
//  AIM
//
//  Created by Oskar Szydlowski on 04.03.2018.
//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
//

#import "Track.h"

@implementation Track

-(instancetype)initWithAttributesDictionary:(NSDictionary<NSString *,NSString *> *)attributes {
    if (attributes[@"title"] == nil) {
        return nil;
    }
    
    if (attributes[@"artist"] == nil) {
        return nil;
    }
    
    if (attributes[@"album"] == nil) {
        return nil;
    }
    
    if (attributes[@"time"] == nil) {
        return nil;
    }
    
    if (attributes[@"duration"] == nil) {
        return nil;
    }
    
    if (attributes[@"imageUrl"] == nil) {
        return nil;
    }
    
    if (attributes[@"status"] == nil) {
        return nil;
    }
    
    if (attributes[@"type"] == nil) {
        return nil;
    }
    
    self = [super init];
    if (self != nil) {
        self.title = attributes[@"title"];
        self.artist = attributes[@"artist"];
        self.album = attributes[@"album"];
        self.playTime = attributes[@"playTime"];
        self.duration = attributes[@"duration"];
        self.imageURL = [[NSURL alloc] initWithString:attributes[@"imageUrl"]];
        self.status = attributes[@"status"];
        self.type = attributes[@"type"];
    }
    return self;
}

@end

File: AIM/AIM/Domain/Entities/Track.swift
//
//  Track.swift
//  AIM
//
//  Created by Oskar Szydlowski on 28.02.2018.
//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
//

import Foundation

enum TrackStatus {
    case history
    case playing
}

enum TrackType {
    case song
}

struct Track {
    
    let title: String
    let artist: String
    let album: String
    let playTime: String
    let duration: String
    let imageURL: URL?
    let status: TrackStatus
    let type: TrackType
}

extension Track {
    
    static func fromAttributeDict(attributeDict: [String: String]) -> Track? {
        guard
            let title = attributeDict["title"],
            let artist = attributeDict["artist"],
            let album = attributeDict["album"],
            let playTime = attributeDict["time"],
            let duration = attributeDict["duration"],
            let imageURLString = attributeDict["imageUrl"],
            let statusString = attributeDict["status"],
            let typeString = attributeDict["type"]
        else {
            assertionFailure("Cannot parse Track data")
            return nil
        }
        
        var status: TrackStatus
        switch statusString {
            case "history":
                status = .history
            case "playing":
                status = .playing
            default:
                assertionFailure("Cannot parse Track data")
                return nil
        }
        
        var type: TrackType
        switch typeString {
            case "song":
                type = .song
            default:
                assertionFailure("Cannot parse Track data")
                return nil
        }
        
        return Track(
            title: title,
            artist: artist,
            album: album,
            playTime: playTime,
            duration: duration,
            imageURL: URL(string: imageURLString),
            status: status,
            type: type
        )
    }
}
